src/main/java/eu/dissco/core/digitalspecimenprocessor/component/HandleComponent.java

```diff
@@ -11,36 +15,60 @@
+import org.springframework.http.HttpStatus;
+import org.springframework.http.HttpStatusCode;
 import org.springframework.stereotype.Component;
 import org.springframework.web.reactive.function.BodyInserters;
 import org.springframework.web.reactive.function.client.WebClient;
+import reactor.core.publisher.Mono;
+import reactor.util.retry.Retry;
 
 @Component
 @RequiredArgsConstructor
 @Slf4j
 public class HandleComponent {
 
-    @Qualifier("handleClient")
-    private final WebClient handleClient;
+  @Qualifier("handleClient")
+  private final WebClient handleClient;
+  private final TokenAuthenticator tokenAuthenticator;
+  private final KafkaPublisherService kafkaService;
 
-    private final TokenAuthenticator tokenAuthenticator;
-
-    public JsonNode postHandle(List<JsonNode> requestBody)throws PidAuthenticationException, PidCreationException {
-        var token = "Bearer " + tokenAuthenticator.getToken();
-        var response = handleClient.post()
-                .uri(uriBuilder -> uriBuilder.path("batch").build())
-                .body(BodyInserters.fromValue(requestBody))
-                .header("Authorization", token)
-                .acceptCharset(StandardCharsets.UTF_8)
-                .retrieve()
-                .bodyToMono(JsonNode.class);
-
-        try {
-            return response.toFuture().get();
-        } catch (InterruptedException | ExecutionException e) {
-            Thread.currentThread().interrupt();
-            log.info(response.toString());
-            throw new PidCreationException("An error has occurred in creating a PID. More information: " + e.getMessage());
-        }
+  public JsonNode postHandle(List<JsonNode> requestBody, List<DigitalSpecimenEvent> digitalSpecimenEvents)
+      throws PidAuthenticationException, PidCreationException, JsonProcessingException {
+    var token = "Bearer " + tokenAuthenticator.getToken();
+    var response = handleClient.post()
+        .uri(uriBuilder -> uriBuilder.path("batch").build())
+        .body(BodyInserters.fromValue(requestBody))
+        .header("Authorization", token)
+        .acceptCharset(StandardCharsets.UTF_8)
+        .retrieve()
+        .onStatus(HttpStatus.UNAUTHORIZED::equals, r -> Mono.error(
+            new PidAuthenticationException("Unable to authenticate with Handle Service.")))
+        .onStatus(HttpStatusCode::is4xxClientError,
+            r -> Mono.error(new PidCreationException("Unable to create PID")))
+        .bodyToMono(JsonNode.class)
+        .retryWhen(Retry.fixedDelay(3, Duration.ofSeconds(2))
+            .filter(WebClientUtils::is5xxServerError)
+            .onRetryExhaustedThrow((retryBackoffSpec, retrySignal) ->
+                new PidCreationException(
+                    "External Service failed to process after max retries")
+            ));
+    try {
+      return response.toFuture().get();
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+    } catch (ExecutionException e) {
```

Comment:
I would do it slightly differently. Keep the HandleComponent only for the call. If it throws an exception let the ProcessingService handle those exceptions. That is the cooridinator of the whole flow.